src/client/datascience/ipywidgets/cdnWidgetScriptSourceProvider.ts

```diff
@@ -81,53 +81,54 @@ export class CDNWidgetScriptSourceProvider implements IWidgetScriptSourceProvide
         // First see if we already have it downloaded.
         const key = this.getModuleKey(moduleName, moduleVersion);
         const diskPath = path.join(this.localResourceUriConverter.rootScriptFolder.fsPath, key, 'index.js');
-        let cached = this.cache.get(key);
         let tempFile: TemporaryFile | undefined;
 
         // Might be on disk, try there first.
-        if (!cached) {
-            if (diskPath && (await this.fs.localFileExists(diskPath))) {
-                const scriptUri = (await this.localResourceUriConverter.asWebviewUri(Uri.file(diskPath))).toString();
-                cached = { moduleName, scriptUri, source: 'cdn' };
-                this.cache.set(key, cached);
-            }
+        if (diskPath && (await this.fs.localFileExists(diskPath))) {
+            const scriptUri = (await this.localResourceUriConverter.asWebviewUri(Uri.file(diskPath))).toString();
+            return { moduleName, scriptUri, source: 'cdn' };
         }
 
         // If still not found, download it.
-        if (!cached) {
-            try {
-                // Make sure the disk path directory exists. We'll be downloading it to there.
-                await this.fs.createLocalDirectory(path.dirname(diskPath));
-
-                // Then get the first one that returns.
-                tempFile = await this.downloadFastestCDN(moduleName, moduleVersion);
-                if (tempFile) {
-                    // Need to copy from the temporary file to our real file (note: VSC filesystem fails to copy so just use straight file system)
-                    await this.fs.copyLocal(tempFile.filePath, diskPath);
-
-                    // Now we can generate the script URI so the local converter doesn't try to copy it.
-                    const scriptUri = (
-                        await this.localResourceUriConverter.asWebviewUri(Uri.file(diskPath))
-                    ).toString();
-                    cached = { moduleName, scriptUri, source: 'cdn' };
-                } else {
-                    cached = { moduleName };
-                }
-            } catch (exc) {
-                traceError('Error downloading from CDN: ', exc);
-                cached = { moduleName };
-            } finally {
-                if (tempFile) {
-                    tempFile.dispose();
-                }
+        try {
+            // Make sure the disk path directory exists. We'll be downloading it to there.
+            await this.fs.createLocalDirectory(path.dirname(diskPath));
+
+            // Then get the first one that returns.
+            tempFile = await this.downloadFastestCDN(moduleName, moduleVersion);
+            if (tempFile) {
+                traceInfo(`Widget Script downloaded for ${moduleName}:${moduleVersion}`);
+                // Need to copy from the temporary file to our real file (note: VSC filesystem fails to copy so just use straight file system)
+                await this.fs.copyLocal(tempFile.filePath, diskPath);
```

Comment:
Previously we'd have two downloads & the second download overwriting the existing file.
While overwriting the existing file, the webview is busy fetching the file from disc (which was downloaded the first time).

Thus we end up corrupting the file (ie writing to a file while its being read by another process)

Solution - ensure we download & copy once (ensure we cache the promise, old cache implementation was not right).

Reply:
FYI - This was a genuine bug (overwriting existing file & corrupting it), not a flaky test.

Reply:
FYI - This was a genuine bug (overwriting existing file & corrupting it), not a flaky test.

Reply:
I'm not seeing the change? Were is the cache of the promise?